src/transformations/InsertFinalNewline.ts

```diff
@@ -23,22 +24,59 @@ export default class InsertFinalNewline extends PreSaveTransformation {
 		const lineCount = doc.lineCount;
 		const lastLine = doc.lineAt(lineCount - 1);
 
-		if (!editorconfigProperties.insert_final_newline
-			|| lineCount === 0
-			|| lastLine.isEmptyOrWhitespace) {
+		if (lineCount === 0) {
 			return { edits: [] };
 		}
 
-		const position = new Position(
-			lastLine.lineNumber,
-			lastLine.text.length
-		);
+		if (editorconfigProperties.insert_final_newline) {
+			if (lastLine.isEmptyOrWhitespace) {
+				return { edits: [] };
+			} else {
+				const position = new Position(
+					lastLine.lineNumber,
+					lastLine.text.length
+				);
 
-		const eol = get(editorconfigProperties, 'end_of_line', 'lf').toUpperCase();
+				const eol = get(editorconfigProperties, 'end_of_line', 'lf').toUpperCase();
 
-		return {
-			edits: [ TextEdit.insert(position, this.lineEndings[eol]) ],
-			message: `insertFinalNewline(${eol})`
-		};
+				return {
+					edits: [ TextEdit.insert(position, this.lineEndings[eol]) ],
+					message: `insertFinalNewline(${eol})`
+				};
+			}
+		} else {
+			if (lastLine.isEmptyOrWhitespace) {
```

Comment:
There's a lot of logic going on in this block. Consider extracting it into a `return deleteFinalNewlines();` call. It should make the rest of the code easier to read as well if we `return insertFinalNewline();` and `return doNothing();` for the "no edits" case.